cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,472 @@
+  }
+
+  private void validateBrokersAllowedReplicaMoveExist(ClusterModel clusterModel,
+                                                      OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    Set<Integer> brokersAllowedReplicaMove = GoalUtils.aliveBrokersNotExcludedForReplicaMove(clusterModel, optimizationOptions);
+    if (brokersAllowedReplicaMove.isEmpty()) {
+      // Handle the case when all alive brokers are excluded from replica moves.
+      throw new OptimizationFailureException("Cannot take any action as all alive brokers are excluded from replica moves.");
+    }
+  }
+
+  /**
+   * Check if requirements of this goal are not violated if this action is applied to the given cluster state,
+   * false otherwise.
+   *
+   * @param clusterModel The state of the cluster.
+   * @param action Action containing information about potential modification to the given cluster model.
+   * @return True if requirements of this goal are not violated if this action is applied to the given cluster state,
+   * false otherwise.
+   */
+  @Override
+  protected boolean selfSatisfied(ClusterModel clusterModel, BalancingAction action) {
+    Broker sourceBroker = clusterModel.broker(action.sourceBrokerId());
+    Replica replicaToBeMoved = sourceBroker.replica(action.topicPartition());
+
+    if (replicaToBeMoved.broker().replica(action.topicPartition()).isCurrentOffline()) {
+      return action.balancingAction() == ActionType.INTER_BROKER_REPLICA_MOVEMENT;
+    }
+    // Moving leader replica from more abundant broker is considered as self-satisfied
+    return sourceBroker.numLeadersFor(replicaToBeMoved.topicPartition().topic()) > minTopicLeadersPerBroker();
+  }
+
+  private boolean actionAffectsRelevantTopics(BalancingAction action) {
+    if (_mustHaveTopicLeadersPerBroker.contains(action.topic())) {
+      return true;
+    }
+    return action.balancingAction() == ActionType.INTER_BROKER_REPLICA_SWAP &&
+           _mustHaveTopicLeadersPerBroker.contains(action.destinationTopic());
+  }
+
+  /**
+   * Update goal state after one round of self-healing / rebalance.
+   *
+   *  @param clusterModel The state of the cluster.
+   * @param optimizationOptions Options to take into account during optimization.
+   */
+  @Override
+  protected void updateGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    // Sanity check: No self-healing eligible replica should remain at a dead broker/disk.
+    GoalUtils.ensureNoOfflineReplicas(clusterModel, name());
+    // Sanity check: No replica should be moved to a broker, which used to host any replica of the same partition on its broken disk.
+    GoalUtils.ensureReplicasMoveOffBrokersWithBadDisks(clusterModel, name());
+    ensureBrokersAllHaveEnoughLeaderReplicaOfTopics(clusterModel, optimizationOptions);
+    clusterModel.brokers().forEach(broker -> broker.untrackSortedReplicas(_replicaSortName));
```

Comment:
The finally block in `AbstractGoal` already calls the following:
```
      // Clear any sorted replicas tracked in the process of optimization.
      clusterModel.clearSortedReplicas();
```

So calling this function here seems redundant. Can we drop this?
-- also note that for such uses we should have preferred `ClusterModel#untrackSortedReplicas(String)`.